accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -1461,3 +1461,319 @@ wrapper class is so huge that it is likely to be a major use case for
+
+A `for (D x in e) body` loop, whether statement or element, with value
+declaration `D x` and `e` having static type, `T`, we require that `T`  is
+assignable to `Iterable<dynamic>`, which we today knows is equivalent to
+implementing `Iterable<S>` for some type `S`, or being `dynamic` or a bottom
+type.
+
+The behavior of the for-in statement is currently specified by a *rewrite* to
+the following code:
+
+```dart
+T _$id1 = e;
+var _$id2 = _$id1.iterator;
+while (_$id2.moveNext()) {
+  D x = _$id2.current;
+  {
+    body  
+  }
+}
+```
+
+That rewrite is no longer valid with extension types. The type `T` may be an
+extension type which *does* implement `Iterable<S>` for some `S`, but which then
+*shadows* the default `iterator`, and returns something entirely unrelated to
+`Iterator<S>`. _We do not intend to invoke that extension type member._
+
+The behavior of a `for`/`in` element is defined in terms of a `for`/`in`
+statement.
+
+##### New specification:
+
+It’s (still) a compile-time error if `T` is not assignable to
+`Iterable<dynamic>`. _This, also still, means that `T` is a bottom type,
+`dynamic` or a type which implements `Iterable<S>` for some `S`. That type may
+just also be an extension type._
+
+If `T` implements `Iterable<S>`, let `E` be `S`, otherwise let `E` be `Never` if
+`T` is a bottom type, or `dynamic` if `T` is `dynamic`.
+
+Now rewrite the `for`-`in` loop to:
+
+```dart
+Iterable<E> _$id1 = e;
+Iterator<E> _$id2 = _$id1.iterator;
+while (_$id2.moveNext()) {
+  D x = _$id2.current;
+  {
+    body
+  }  
+}
+```
+
+Like before, any errors that would occur in that program are reported as errors
+in the original `for`-`in` statement.
+
+_**TODO**: Specify this without using a “desugaring” rewrite. Problems like
+these are exactly why rewrites are problematic, they tend to introduce *more*
+information, through the extra syntax or inferred types, than what the original
+program warranted. Also, the rewrite supposedly happens after type inference, so
+we need to either perform type inference on the desugared code, or assign a
+context type and static type to each new expression, because those are
+referenced by the dynamic semantics._ 
+
+##### Consequences to existing code
+
+The only real change is that the type of `_$id1` is not the static type of `e`.
+That means that if there is a class which overrides its iterator to be more
+specific than required, that information is lost.
+
+Example:
+
+```dart
+class VeryInts extends Iterable<num> {
+  Iterator<int> get iterator => <int>[1].iterator;
+}
+void main() {
+  for (int i in VeryInts()) print(i);
+}
+```
+
+With the previous *specification*, the type that `.iterator` should be accessed
+on was `VeryInts`, which means that `_$id2` will get static type
+`Iterator<int>`, and its `.current` is then assignable to `int i`. With the new
+specification, we don’t look at the members of the actual type, instead we
+immediately up-cast to `Iterable<E>` and work with that.
```

Comment:
I'd definitely prefer that the developer must write that upcast and the treatment after that point is completely standard, rather than getting into a situation where some members are called based on the type `Iterable<E>` for some `E` while the receiver has an extension type `V` which has an `iterator` member with a different behavior.